### ISSUE TITLE
`<Esc>` from mapping executes cmdline in command mode

### ISSUE THREAD

**Issue by tmillr** [neovim/neovim#21585]
### Neovim version (nvim -v)

NVIM v0.9.0-dev-1763+g9be24110f-dirty (Build type: Release)

### Vim (not Nvim) behaves the same?

unsure (probably?)

### Operating system/version

macOS 11.7.3

### Terminal name/version

kitty 0.26.5

### $TERM environment variable

xterm-kitty

### Installation

`brew install --HEAD neovim`

### How to reproduce the issue

Within an interactive terminal/shell, do:

```vim
nvim --clean -c "cnoremap <expr> <Esc> '<Esc>'"
```

Now, type `:cq` followed by the escape key on your keyboard.

### Expected behavior

I expect the behavior to be identical to what occurs when escape is pressed while not mapped. That is:

1. whatever was typed on the cmdline is saved in the command history
2. the cmdline itself should clear/become blank (no text)
3. command mode is exited without any commands being executed (unless there are `CmdlineLeave` autocmds defined, of course)

TLDR: the cmdline is "aborted"

### Actual behavior

The difference in actual vs expected behavior occurs when escape is pressed, wherein the current contents of the cmdline are executed causing Neovim to exit with exit-code `1` (due to the `cq` command being executed). Generally speaking, Neovim behaves as if `<Esc>` was mapped to `<CR>`, making it very difficult, if not impossible, to have a useful command mode mapping for the escape key (e.g. a conditional mapping such as: _close the completion window if it is visible, otherwise abort the cmdline with `<Esc>` as if it had never been mapped in the first place_).

In other words, Neovim essentially behaves as if the `x` option/flag of `&cpo` (aka `cpoptions`) is set, even though this is definitely not the case (as `x` is off by default, plus I continue to experience this bug even after explicitly running `set cpo-=x`). Neovim does **not** behave this way when escape is pressed while not mapped however, so I don't quite understand why the behavior differs here (receiving escape naturally vs via <expr> map). I am unsure if this is a bug, user error, or expected behavior, etc.

**Comment by zeertzjq** [comment]
This is expected behavior.

**Comment by bew** [comment]
@zeertzjq Could you explain why this behavior would be wanted, expected?

**Comment by tmillr** [comment]
> This is expected behavior.

Thanks for the quick answer. Is there a reason for this? Because I don't understand why cpo's `x` wouldn't be honored. Is there a workaround?

**Comment by zeertzjq** [comment]
Use `<C-C>`

**Comment by zeertzjq** [comment]
Hmm, it seems that hardly anyone knows what is the reason for this. Let's reopen this.

Ref: https://groups.google.com/g/vim_use/c/8Mhs9spyzCM/m/qEFr6AFshWcJ

**Comment by tmillr** [comment]
> Hmm, it seems that hardly anyone knows what is the reason for this. Let's reopen this.
> 
> Ref: https://groups.google.com/g/vim_use/c/8Mhs9spyzCM/m/qEFr6AFshWcJ

Nice find! Interestingly, this behavior even appears to be documented in the proper location (and in an obvious one, I can't believe I didn't think of doing `:h c_<Esc>`).

If I had to guess, maybe it was done as some kind of workaround or something, or done in order to avoid some bug/issue. It would certainly seem like an odd preference given how unintuitive it is, especially considering how:

- it completely ignores the absence of the `x` flag in `:help 'cpoptions'`
- the `x` flag is absent by default

> **P.S:** I found #15518 when I was searching for pre-existing issues...idk, but maybe it is related in some way.

**Comment by JosefLitos** [comment]
Is there a possibility this behaviour will get changed to the expected one?

**Comment by scienceplease** [comment]
> maybe it was done as some kind of workaround or something, or done in order to avoid some (other) bug/issue. It would certainly seem like an odd preference given just how unintuitive it is, and especially considering how:

You can just map `<Esc>` to `<C-c>` for command mode to get the expected behavior.

**Comment by scienceplease** [comment]
> Is there a possibility this behaviour will get changed to the expected one?

You can just map <Esc> to <C-c> for command mode to get the expected behavior.

**Comment by scienceplease** [comment]
> Is there a possibility this behaviour will get changed to the expected one?

You can just map `<Esc>` to `<C-c>` for command mode to get the expected behavior.

**Comment by JosefLitos** [comment]
Oh, I just noticed it already does work as expected by default (at least for me). It only changes behaviour after loading `nvim-cmp` because my cmp config modifies `Esc` in command mode. However, if it does work by default, it is weird this issue is still opened. I must be missing something.

**Comment by tmillr** [comment]
> However, if it does work by default, it is weird this issue is still opened. I must be missing something.

Did you follow all of the instructions under `How to reproduce the issue` above?

**Comment by JosefLitos** [comment]
> Did you follow all of the instructions under `How to reproduce the issue` above?

Oh, it's the use in a mapping that causes it, I see. That's the reason why it acts like that with cmp too.
Thanks
